core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -319,51 +319,68 @@ class BlockManagerMasterEndpoint(
-                val blocks = blocksToDeleteByShuffleService.getOrElseUpdate(mapStatus.location,
-                  new mutable.HashSet[BlockId])
-                blocks ++= blocksToDel
+      val shuffleClient = externalBlockStoreClient.get
+      // Find all shuffle blocks on executors that are no longer running
+      val blocksToDelete = new mutable.HashMap[BlockManagerId, mutable.HashSet[BlockId]]
+      mapOutputTracker.shuffleStatuses.get(shuffleId) match {
+        case Some(shuffleStatus) =>
+          shuffleStatus.withMapStatuses { mapStatuses =>
+            mapStatuses.foreach { mapStatus =>
+              // Check if the executor has been deallocated
+              if (!blockManagerIdByExecutor.contains(mapStatus.location.executorId)) {
+                val blocksToDel = shuffleManager.shuffleBlockResolver
+                  .getBlocksForShuffle(shuffleId, mapStatus.mapId)
+                if (blocksToDel.nonEmpty) {
+                  val blocks = blocksToDelete.getOrElseUpdate(mapStatus.location,
+                    new mutable.HashSet[BlockId])
+                  blocks ++= blocksToDel
+                }
               }
             }
           }
-        }
+          removeShuffleFromShuffleServicesFutures ++= blocksToDelete.map { case (bmId, blockIds) =>
+            Future[Boolean] {
+              val numRemovedBlocks = shuffleClient.removeBlocks(
+                bmId.host,
+                bmId.port,
+                bmId.executorId,
+                blockIds.map(_.toString).toArray)
+              numRemovedBlocks.get(defaultRpcTimeout.duration.toSeconds,
+                TimeUnit.SECONDS) == blockIds.size
+            }
+          }.toSeq
+        case None =>
+          logDebug(s"Asked to remove shuffle blocks from " +
+            s"shuffle service for unknown shuffle ${shuffleId}")
       }
     }
-
-    val removeShuffleFromShuffleServicesFutures =
-      externalBlockStoreClient.map { shuffleClient =>
-        blocksToDeleteByShuffleService.map { case (bmId, blockIds) =>
-          Future[Boolean] {
-            val numRemovedBlocks = shuffleClient.removeBlocks(
-              bmId.host,
-              bmId.port,
-              bmId.executorId,
-              blockIds.map(_.toString).toArray)
-            numRemovedBlocks.get(defaultRpcTimeout.duration.toSeconds,
-              TimeUnit.SECONDS) == blockIds.size
-          }
-        }
-      }.getOrElse(Seq.empty)
-
+    // It needs to be invoked at last to avoid cleaning up shuffleStatuses in mapOutputTracker
+    // too early before used by [[removeShuffleFromShuffleServicesFutures]]
+    val removeMsg = RemoveShuffle(shuffleId)
+    val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
+      bm.storageEndpoint.ask[Boolean](removeMsg).recover {
+        // use false as default value means no shuffle data were removed
+        handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
```

Comment:
I just move `removeShuffleFromExecutorsFutures` to the last.
It needs to be invoked at last to avoid cleaning up shuffleStatuses in mapOutputTracker to early. Otherwise `mapOutputTracker.shuffleStatuses.get(shuffleId)` may be none sometimes
Please refer to [unregisterShuffle  codes](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/storage/BlockManagerStorageEndpoint.scala#L59)